### PR TITLE
WIP User authentication security

### PR DESCRIPTION
**Pull request by Dpalazzari**
@Laszlo-JFLMTCO @drod1000 

We did both Authentication cards one PR.

The cards we solved:

Background: An authenticated user
As an Authenticated User
I cannot view another user's private data (current or past orders, etc)
I cannot view the administrator screens or use admin functionality
I cannot make myself an admin

Background: An unauthenticated user and their abilities
As an Unauthenticated User
I cannot view another user's private data, such as current order, etc.
I should be redirected to login/create account when I try to check out.
I cannot view the administrator screens or use administrator functionality.
I cannot make myself an administrator.

We also have pretty much completed the first half of the checking out card.

Note for person/people who do checking out card: When you click on checkout you are sent to the orders#new method which checks to see if you are a authenticated user or not. If you are it redirects you to checkout_path. The idea is, from there, you would take the cart items in the current session and create an order and change the order status to 'ordered'.

`get '/orders' => 'orders#create', :as "Checkout"` 

Feel free to change if you don't like it or find a better design. 🤖

### CODE REVIEW

**Comment by Dpalazzari** [comment]
Going to finish issue #23 as well

**Comment by Dpalazzari** [comment]
Also, after merging, make sure you bundle because I added a new gem to the gemfile.

File: app/controllers/admin/base_controller.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Admin::BaseController < ApplicationController
+
+  before_action :require_admin
+
+  def require_admin
+    render file: "/public/404" unless current_user && current_user.admin?
```
Comment (Laszlo-JFLMTCO) [inline app/controllers/admin/base_controller.rb:6]: @Dpalazzari @njgheorghita Do we have any specific message on the 404 page to let the user know what happened? Or is it intentional not to even give the impression that the page exists for some users? I could see the benefit, just wanted to confirm the thinking behind this implementation. The alternative I have in mind is flash message and redirect to `root_path`. FYI @drod1000

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -16,7 +16,11 @@ def create
   end
 
   def show
-    @user = User.find(session[:user_id])
+    if session[:user_id].nil?
+      render file: "/public/404"
```
Comment (Laszlo-JFLMTCO) [inline app/controllers/users_controller.rb:20]: @Dpalazzari @njgheorghita Wouldn't a redirect to `root_path` with a flash message to the user would be more user friendly for this use case? Can you please help to confirm why showing 404 page is done here? FYI @drod1000

File: config/environments/test.rb [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Rails.application.configure do
   # Settings specified here will take precedence over those in config/application.rb.
-
+  config.middleware.use RackSessionAccess::Middleware
```
Comment (Laszlo-JFLMTCO) [inline config/environments/test.rb:3]: @drod1000 @njgheorghita Is my understanding correct that this is necessary to support testing of the various user roles? FYI @drod1000

**Comment by Laszlo-JFLMTCO** [comment]
Great job @Dpalazzari and @njgheorghita !!! I will pull this branch and run locally, but initial review of the changes are looking good!!! I will switch to checkout user story shortly. Thank you for setting it up for an easier implementation! FYI @drod1000

File: app/controllers/admin/base_controller.rb (hunk 1 above)
Comment (drod1000) [reply app/controllers/admin/base_controller.rb:6]: @Dpalazzari @njgheorghita I was wondering the same thing here. I was thinking a 403 might be better or the root and forbidden message like @Laszlo-JFLMTCO  mentioned

**Comment by drod1000** [comment]
@Dpalazzari @njgheorghita I pulled this branch down and implemented logout functionality on a dummy branch and I can confirm that it's working as expected. Great job guys!

File: app/controllers/admin/base_controller.rb (hunk 1 above)
Comment (njgheorghita) [reply app/controllers/admin/base_controller.rb:6]: no specific message on the 404 page to let the user know what happened. 
the 404 message displays with the navbar above it, so i'm pretty sure a flash message will display if we create one, without having to change the redirect path. 
though, i'm not against redirecting to root if that's what guys think is better

File: config/environments/test.rb (hunk 3 above)
Comment (njgheorghita) [reply config/environments/test.rb:3]: exactly, it's for .set_rack_session() in the feature tests to login a particular user

**Pull request merged**